specs/ExecuteScript.md

```diff
@@ -0,0 +1,304 @@
+# Background
+Currently WebView2 supports script execution in the main page only -
+documentation can be found [here](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2?view=webview2-1.0.902.49#executescript).
+With introduction of API for [separate iframes](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2frame?view=webview2-1.0.902.49) we find it useful to support script execution in child frames also.
+The time when script can be executed is tightly coupled with the navigation state of the page,
+therefore we additionally providing [navigation events](https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/navigation-events) in the iframes - NavigationStarting, ContentLoading, NavigationCompleted and DOMContentLoaded event.
+
+In this document we describe API additions to support ExecuteScript and navigation
+events in iframes. We'd appreciate your feedback.
+
+
+# Description
+Using ExecuteScript user can run script in the context of a calling frame and receive the script result in the callback if provided.
+If ExecuteScript is run during the navigation it should be called after ContentLoading event.
+
+# Examples
+## ExecuteScript
+
+## .NET, WinRT
+```c#
+async void InjectScriptIFrameCmdExecuted(object target, ExecutedRoutedEventArgs e)
```

Comment:
Can we come up with sample code that would demonstrate a real sort of scenario that an end developer would want to do? In this case this sample code is more for exploring the APIs which is very useful in our sample app, but doesn't give an example of why an end developer would want to use the API or how to use the things together.

Reply:
Agree on showing more real world scenario. I like our tests use-cases better. For example executing script on content loading:

```
webview2->add_FrameCreated(
        Microsoft::WRL::Callback<
            ICoreWebView2FrameCreatedEventHandler>([this, &iframe_url](
                                                       ICoreWebView2* webview,
                                                       ICoreWebView2FrameCreatedEventArgs*
                                                           args) -> HRESULT {
          Microsoft::WRL::ComPtr<ICoreWebView2Frame> webviewFrame;
          EXPECT_EQ(S_OK, args->get_Frame(&webviewFrame));

          webviewFrame->add_ContentLoading(
              Microsoft::WRL::Callback<
                  ICoreWebView2ExperimentalFrameContentLoadingEventHandler>(
                  [&iframe_url](
                      ICoreWebView2Frame* frame,
                      ICoreWebView2ContentLoadingEventArgs* args) -> HRESULT {
                    frame->ExecuteScript(
                        L"window.location.href.toString()",
                        Microsoft::WRL::Callback<
                            ICoreWebView2ExecuteScriptCompletedHandler>(
                            [&iframe_url](HRESULT error,
                                          LPCWSTR result) -> HRESULT {
                              EXPECT_EQ(error, S_OK);
                              const std::wstring expected_response =
                                  L"\"" + iframe_url + L"\"";
                              EXPECT_EQ(expected_response, result);
                              return S_OK;
                            })
                            .Get());

                    return S_OK;
                  })
                  .Get(),
              NULL);

          return S_OK;
        }).Get(),
        nullptr);
```
Do you think it can work here as sample?

Reply:
Something along these lines seems good

Reply:
Updated cpp and .NET samples.